gk/fib.c

```diff
@@ -0,0 +1,1413 @@
+
+		if (gt_addr != NULL || gw_addr != NULL)
+			return -1;
+
+		ip_prefix_fib = init_drop_fib_locked(prefix, gk_conf);
+		if (ip_prefix_fib == NULL)
+			return -1;
+
+		break;
+	}
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"Invalid fib action %u at %s\n", action, __func__);
+		return -1;
+		break;
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: add a fib entry for ip prefix = %s with action = %u]\n",
+		prefix->str, action);
+
+	return 0;
+}
+
+/* Return 0 when @gw_addr is not included in @prefix; otherwise return -1. */
+static int
+check_gateway_prefix(struct ip_prefix *prefix, struct ipaddr *gw_addr)
+{
+	if (gw_addr->proto == ETHER_TYPE_IPv4) {
+		uint32_t ip4_mask =
+			rte_cpu_to_be_32(~0ULL << (32 - prefix->len));
+		if ((prefix->addr.ip.v4.s_addr ^
+				gw_addr->ip.v4.s_addr) & ip4_mask)
+			return 0;
+	} else if (likely(gw_addr->proto == ETHER_TYPE_IPv6)) {
+		uint64_t ip6_mask;
+		uint64_t *pf = (uint64_t *)prefix->addr.ip.v6.s6_addr;
+		uint64_t *gw = (uint64_t *)gw_addr->ip.v6.s6_addr;
+
+		if (prefix->len == 0)
+			return -1;
+		else if (prefix->len <= 64) {
+			ip6_mask = rte_cpu_to_be_64(
+				~0ULL << (64 - prefix->len));
+			if ((pf[0] ^ gw[0]) & ip6_mask)
+				return 0;
+		} else {
+			ip6_mask = rte_cpu_to_be_64(
+				~0ULL << (128 - prefix->len));
+			if ((pf[0] != gw[0]) ||
+					((pf[1] ^ gw[1]) & ip6_mask)) {
+				return 0;
+			}
+		}
+	} else
+		rte_panic("Unexpected condition at %s: unknown IP type %hu\n",
+			__func__, gw_addr->proto);
+
+	return -1;
+}
+
+/*
+ * This function makes sure that only a drop or another grantor entry
+ * must be able to be longer than a grantor or a drop prefix.
+ *
+ * TODO A potential bug needs to be fixed here.
+ * Assuming there's prefix 10.1.1.0/24 that leads to a gateway and
+ * the prefix being added is 10.1/16. That is, the gateway entry would
+ * be a security hole.
+ *
+ * The solution is to extend RTE LPM table to have an iterator over
+ * its entries. The iterator should require a prefix as parameter and
+ * list all entries as long as a given prefix or longer.
+ * To list all entries, the prefix would have length zero.
+ */
+static int
+check_prefix_locked(struct ip_prefix *prefix,
+	enum gk_fib_action action, struct gk_config *gk_conf)
+{
+	uint8_t i;
+	int ip_prefix_present;
+	struct gk_fib *ip_prefix_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (action == GK_DROP || action == GK_FWD_GRANTOR)
+		return 0;
+
+	if (prefix->addr.proto == ETHER_TYPE_IPv4) {
+		for (i = 0; i < prefix->len; i++) {
+			uint32_t fib_id;
+			ip_prefix_present = rte_lpm_is_rule_present(
+				ltbl->lpm, ntohl(prefix->addr.ip.v4.s_addr),
+				prefix->len, &fib_id);
+			if (!ip_prefix_present)
+				continue;
+
+			ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+			if (ip_prefix_fib->action != GK_FWD_GRANTOR &&
+					ip_prefix_fib->action != GK_DROP) {
+				return -1;
+			}
```

Comment:
In my understanding of the style rule, it would be okay to leave the braces if the single statement took multiple lines. But the condition has no impact on the braces; parentheses already enclose the condition.